src/misc/messages.cpp

```diff
@@ -62,56 +84,79 @@ class Message {
 		if (rendered_msg.empty())
 			rendered_msg = convert_ansi_markup(markup_msg.c_str());
 		assert(rendered_msg.length());
-		return rendered_msg.c_str();
+
+		if (!is_utf8_msg)
+			return rendered_msg.c_str();
+
+		const uint16_t cp = UTF8_GetCodePage();
+		if (rendered_cp_msg[cp].empty()) {
+			if (!UTF8_RenderForDos(rendered_msg, rendered_cp_msg[cp], cp))
+				LOG_WARNING("LANG: Problem rendering string");
+			assert(rendered_cp_msg[cp].length());
+		}
+		return rendered_cp_msg[cp].c_str();
 	}
 
-	void Set(const char *markup)
+	void Set(const char *markup, const bool is_utf8)
 	{
 		assert(markup);
-		markup_msg = markup;
+		markup_msg  = markup;
+		is_utf8_msg = is_utf8;
+
 		rendered_msg.clear();
+		markup_cp_msg.clear();
+		rendered_cp_msg.clear();
 	}
 };
 
 static std::unordered_map<std::string, Message> messages;
 static std::deque<std::string> messages_order;
 
-// Add the message if it doesn't exist yet
-void MSG_Add(const char *name, const char *markup_msg)
+static void msg_add(const char *name, const char *markup_msg, const bool is_utf8)
 {
-	const auto &pair = messages.try_emplace(name, markup_msg);
+	const auto &pair = messages.try_emplace(name, markup_msg, is_utf8);
 	if (pair.second) // if the insertion was successful
 		messages_order.emplace_back(name);
 }
 
+// Add the message if it doesn't exist yet
+void MSG_Add(const char *name, const char *markup_msg)
+{
+	msg_add(name, markup_msg, false);
```

Comment:
I think I mentioned this before: at the very least do this:

```cpp
constexpr auto is_utf8 = false;
msg_add(name, markup_msg, is_utf8);
```

But ditching the bool flag and having two variants of each function would be preferable:

```cpp
msg_add(name, markup_msg);
msg_add_utf8(name, markup_msg);
```

Internally, it would still look like this, of course — the public API and how things look at the call sites is the important thing here:

```cpp
static void _msg_add(const char *name, const char *markup_msg, const bool is_utf8);

void msg_add(const char *name, const char *markup_msg)
{
    constexpr auto  is_utf8 = false;
    _msg_add(name, markup_msg, is_utf8);
}

void msg_add_utf8(const char *name, const char *markup_msg)
{
    constexpr auto  is_utf8 = true;
    _msg_add(name, markup_msg, is_utf8);
}
```

My personal view is that boolen arguments should almost never be used in multi-argument functions. Ideally you'd only use a boolean arg when it's the only argument, in most other cases I consider it a code-smell and readability issue, especially in languages like C++ that has no support for named arguments (e.g. this way of calling it would be perfectly acceptable: `msg_add(name, markup_msg, is_utf8 = true`)

Reply:
Since 'msg_add' is only called twice in the source code (once for UTF-8, and once for non UTF-8 string), I have simply added the 'constexpr auto is_utf8' before each call.

BTW, I intend to make all the calls an UTF-8 ones in the future, but I feel this would have been too much for a single PR.